src/commands/filter_parser.cc

```diff
@@ -811,23 +836,20 @@ absl::StatusOr<std::unique_ptr<query::Predicate>> FilterParser::ParseExpression(
             absl::StrCat("Expected ')' after expression got '",
                          expression_.substr(pos_, 1), "'. Position: ", pos_));
       }
-      if (prev_predicate) {
-        node_count_++;  // Count the ComposedPredicate Node
-      }
-      prev_predicate =
-          WrapPredicate(std::move(prev_predicate), std::move(predicate), negate,
-                        query::LogicalOperator::kAnd);
+      predicate = MayNegatePredicate(std::move(predicate), negate);
+      and_terms.push_back(std::move(predicate));
     } else if (Match('|')) {
```

Comment:
I did not follow why changes were needed in this function to this extent of tracking a vector of AND and ORs and building it in a final step. We can discuss this. If a major change is really needed, we can discuss how to make it fit the pattern of recursive descent for maintainability